lib/Common/Memory/SectionAllocWrapper.cpp

```diff
@@ -0,0 +1,949 @@
+{
+    uint index = GetNodeIndex(address);
+
+    Node * node = list;
+    while (node != nullptr)
+    {
+        if (node->nodeIndex == index)
+        {
+            return node;
+        }
+
+        node = node->next;
+    }
+
+    return nullptr;
+}
+
+#endif //TARGET_64
+
+static const uint SectionAlignment = 65536;
+
+LPVOID
+AllocLocalView(HANDLE sectionHandle, LPVOID remoteBaseAddr, LPVOID remoteRequestAddress, size_t requestSize)
+{
+    const size_t offset = (uintptr_t)remoteRequestAddress - (uintptr_t)remoteBaseAddr;
+    LARGE_INTEGER mapOffset = { 0 };
+    const size_t offsetAlignment = offset % SectionAlignment;
+#if TARGET_32
+    mapOffset.LowPart = offset - offsetAlignment;
+#elif TARGET_64
+    mapOffset.QuadPart = offset - offsetAlignment;
+#else
+    CompileAssert(UNREACHED);
+#endif
+    LPVOID address = nullptr;
+    SIZE_T viewSize = requestSize + offsetAlignment;
+    int status = NtdllLibrary::Instance->MapViewOfSection(sectionHandle, GetCurrentProcess(), &address, NULL, NULL, &mapOffset, &viewSize, NtdllLibrary::ViewUnmap, NULL, PAGE_READWRITE);
+    if (status != 0 || address == nullptr)
+    {
+        return nullptr;
+    }
+    if (viewSize < requestSize)
+    {
+        NtdllLibrary::Instance->UnmapViewOfSection(GetCurrentProcess(), address);
+        return nullptr;
+    }
+    return (LPVOID)((uintptr_t)address + offsetAlignment);
+}
+
+BOOL
+FreeLocalView(LPVOID lpAddress)
+{
+    const size_t alignment = (uintptr_t)lpAddress % SectionAlignment;
+    NtdllLibrary::Instance->UnmapViewOfSection(GetCurrentProcess(), (LPVOID)((uintptr_t)lpAddress - alignment));
+    return TRUE;
+}
+
+SectionAllocWrapper::SectionAllocWrapper(HANDLE process) :
+    process(process)
+{
+    Assert(process != GetCurrentProcess()); // only use sections when OOP
+}
+
+LPVOID
+SectionAllocWrapper::Alloc(LPVOID requestAddress, size_t dwSize, DWORD allocationType, DWORD protectFlags, bool isCustomHeapAllocation)
+{
+    Assert(isCustomHeapAllocation);
+
+    LPVOID address = nullptr;
+
+#if defined(ENABLE_JIT_CLAMP)
+    // REVIEW: is this needed?
+    AutoEnableDynamicCodeGen enableCodeGen(true);
+#endif
+
+    // for new allocations, create new section and fully map it (reserved) into runtime process
+    if (requestAddress == nullptr)
+    {
+        NtdllLibrary::OBJECT_ATTRIBUTES attr;
+        NtdllLibrary::Instance->InitializeObjectAttributes(&attr, NULL, NtdllLibrary::OBJ_KERNEL_HANDLE, NULL, NULL);
+        LARGE_INTEGER size = { 0 };
+#if TARGET_32
+        size.LowPart = dwSize;
+#elif TARGET_64
+        size.QuadPart = dwSize;
+#endif
+        HANDLE sectionHandle = nullptr;
+        const ULONG allocAttributes = ((allocationType & MEM_COMMIT) == MEM_COMMIT) ? SEC_COMMIT : SEC_RESERVE;
+        int status = NtdllLibrary::Instance->CreateSection(&sectionHandle, SECTION_MAP_READ | SECTION_MAP_WRITE | SECTION_QUERY | SECTION_MAP_EXECUTE, &attr, &size, PAGE_EXECUTE_READWRITE, allocAttributes, NULL);
```

Comment:
>PAGE_EXECUTE_READWRITE [](start = 165, length = 22)

will this make it executable on JIT server, or is this necessary to be executable here since we specify PAGE_EXECUTE while MapViewOfSection?